weave/monitoring2/openai.py

```diff
@@ -0,0 +1,401 @@
+        else:
+            raise ValueError("stream_key must be of the form 'entity/project/stream_name' or 'project/stream_name'")
+
+        streamtable = StreamTable(stream_name, project_name=project_name, entity_name=entity_name)
+        return cls(streamtable)
+
+    def before_send_request(self, context: Context, *args, **kwargs):
+        sig = match_signature(old_create, *args, **kwargs)
+        context.inputs = ChatCompletionRequest.model_validate(sig)
+
+    def before_end(self, context: Context, *args, **kwargs):
+        inputs: ChatCompletionRequest = context.inputs
+        outputs: ChatCompletion = context.outputs
+
+        d = {}
+        if inputs:
+            d["inputs"] = inputs.model_dump()
+        if outputs:
+            d["outputs"] = outputs.model_dump()
+
+        self._streamtable.log(d)
+        self._streamtable.finish()
+
+
+class AsyncChatCompletions:
+    def __init__(self, base_create, callbacks: List[Callback] = None):
+        self._base_create = base_create
+
+        self.callbacks = coalesce(callbacks, [])
+        self.callbacks += [
+            ReassembleStream(),
+            InfoMsg(),
+            LogToStreamTable.from_stream_name("test-stream3", "test-project", "megatruong"),
+        ]
+
+    async def create(self, *args, **kwargs):
+        self.context = Context()
+        if kwargs.get("stream", False):
+            return self._streaming_create(*args, **kwargs)
+        return await self._create(*args, **kwargs)
+
+    async def _create(self, *args, **kwargs):
+        for callback in self.callbacks:
+            await self._use_callback(callback.before_send_request, self.context, *args, **kwargs)
+
+        result = await self._base_create(*args, **kwargs)
+        self.context.outputs = result
+
+        for callback in self.callbacks:
+            await self._use_callback(callback.before_end, self.context, *args, **kwargs)
+
+        return result
+
+    async def _streaming_create(self, *args, **kwargs):
+        for callback in self.callbacks:
+            await self._use_callback(callback.before_send_request, self.context, *args, **kwargs)
+
+        stream = await self._base_create(*args, **kwargs)
+        async for chunk in stream:
+            for callback in self.callbacks:
+                await self._use_callback(callback.before_yield_chunk, self.context, *args, **kwargs)
+
+            yield chunk
+
+            for callback in self.callbacks:
+                await self._use_callback(callback.after_yield_chunk, self.context, *args, **kwargs)
+
+        for callback in self.callbacks:
+            await self._use_callback(callback.before_end, self.context, *args, **kwargs)
+
+    @staticmethod
+    async def _use_callback(f, context, *args, **kwargs):
+        if asyncio.iscoroutinefunction(f):
+            await f(context, *args, **kwargs)
+        else:
+            f(context, *args, **kwargs)
+
+
+class ChatCompletions:
```

Comment:
On both classes, i'd like to see a lot more error handling. If something goes wrong in the patched code, we should catch that and not necessarily crash the user script.